[pfe-accordion]: 1:1

brianferry opened the pull request:
This is the pr for the pfe 1:1 between [pfe4](https://patternfly-react-main.surge.sh/components/accordion/) and [patternfly elements](https://patternflyelements.org/components/accordion/)

bennypowers commented:
make sure to add a changeset and keep it up to date with api changes

marionnegp reviewed (changes requested):
Left a few notes, but this is lookin good!

A couple other things:

- [ ] PF only has a large version of the bordered accordion, but `pfe-accordion` is showing it without borders.
- [ ] Do we need nested accordions in PFE? I know it's something that we have in RHDS, but I don't see an example of nested accordions in PF.

brianferry commented on elements/pfe-accordion/BaseAccordionHeader.ts:
```diff
@@ -0,0 +1,166 @@
+            </span>
+            `}
+          <pfe-icon
+              icon="web-icon-caret-thin-right"
```
So I went and changed the small arrow icon to 10px x 10px and the large icon to 12px x 12px.  Can you tell me if that looks better?

marionnegp replied:
Yep, that looks better! I'm guessing the icons will look different just because of the different PF and PFE icon sets?

brianferry commented on elements/pfe-accordion/pfe-accordion-header.scss:
```diff
@@ -1,246 +1,15 @@
+  --pf-c-accordion__toggle--PaddingRight: var(--pf-global--spacer--md, 1rem);
+  --pf-c-accordion__toggle--PaddingBottom: var(--pf-global--spacer--md, 1rem);
+  --pf-c-accordion__toggle--PaddingLeft: var(--pf-global--spacer--lg, 1.5rem);
+  --pf-c-accordion__toggle--FontFamily:  var(--pf-global--FontFamily--heading--sans-serif,  "RedHatTextUpdated", "Overpass", overpass, helvetica, arial, sans-serif);
```
Fixed

brianferry commented:
@marionnegp - 

- Updated the demos to match the pf demos
- Added the bordered styles to the large variant

We don't need the demo showing the nested accordion though technically it's still possible with how it's set up on our and patternfly's side to have these nested accordions.  Maybe we just don't demo it & it remains an unsupported piece of pfe but rhds it'll be more front and center.  What do you think?

marionnegp commented:
@brianferry, the large heading isn't loading in the bordered demo anymore. Any idea why?
 
<img width="993" alt="Screen Shot 2022-09-19 at 9 13 38 AM" src="https://user-images.githubusercontent.com/95588923/191027143-2cf6b724-8663-44f8-8bed-ae7730e2f3c7.png">

bennypowers commented:
Notes from Design Systems office hours:


User expands a collapsed panel in a `single`, nested accordion
**EXPECTED**: only the parent accordion closes it's open panels
**ACTUAL**: All accordions in the nested accordion's tree close open panels, i.e. the ancestors

1. have every accordion be a single context provider
2. have every header be a single context consumer
3. attach the header's single context provider to the change event
4. in accordion, when handling change events, only act if
   `event.provider === this`, meaning that the accordion object handling the event
   is truly the nearest ancestor

## Potential gotchas:
- can we use `closest`? tagName, shadow dom boundaries
- test for late-upgrading intermediate child (see AB tests on RHDC)

marionnegp commented:
@brianferry, let's keep the nested accordion demo. It's being used somewhere in Customer Portal's Case Management tool.

brianferry commented:
> @brianferry, let's keep the nested accordion demo. It's being used somewhere in Customer Portal's Case Management tool.

Would they be switching over to the rhds accordion when that one is implemented?

marionnegp commented:
@brianferry, I'm not sure if they'll be switching. I misread her message, and they're actually using **PF** for their nested accordion. I guess it does exist in PF.

brianferry commented:
> @brianferry, I'm not sure if they'll be switching. I misread her message, and they're actually using **PF** for their nested accordion. I guess it does exist in PF.

@marionnegp - Font size should be fixed on the demo for large components now.

bennypowers commented on elements/pfe-accordion/BaseAccordionHeader.scss:
```diff
@@ -0,0 +1,135 @@
+:host { 
+    --pf-c-accordion__toggle--PaddingTop: var(--pf-global--spacer--sm, 0.5rem);
```
Also, and I haven't thought this through entirely so disregard if you have a position on this, but we should perhaps move most of these to pfe-accordion.scss, especially if RHDS is going to override them.

brianferry replied:
So you're thinking like instead of setting the var on host in the scss we would do something like 

```css
padding: var(--pf-global--spacer--sm, 0.5rem) var(--pf-global--spacer--md, 1rem) v var(--pf-global--spacer--sm, 0.5rem) var(--pf-global--spacer--md, 1rem);
```

or maybe something like 

```css
padding: var(--pf-c-accordion__toggle--PaddingTop, var(--pf-global--spacer--sm, 0.5rem)) var(--pf-c-accordion__toggle--PaddingRight, var(--pf-global--spacer--md, 1rem)) var(--pf-c-accordion__toggle--PaddingBottom, var(--pf-global--spacer--sm, 0.5rem)) var(--pf-c-accordion__toggle--PaddingLeft, var(--pf-global--spacer--md, 1rem));

```

Where we don't automatically set the variables but like you said we build off of the theming and then they can overwrite either all through the --pf-global styles or individually through the variables

bennypowers replied:
@markcaron do you know if these vars are meant to be user accessible? Is it ok if the user's preferences for these values from higher up the DOM tree are ignored?

Once this question is resolved, i'm good to merge

bennypowers commented on elements/pfe-accordion/BaseAccordionPanel.ts:
```diff
@@ -0,0 +1,47 @@
+
+export class BaseAccordionPanel extends LitElement {
+  static readonly styles = [
+    ...PfeCollapsePanel.styles,
```
we're going to remove collapse, so maybe now's the best time to move those styles over. We can defer this until after but doing the work here will make it a drop easier.

bennypowers commented:
@brianferry thinking it over i'm not certain we need context.

In the change handler, you could do something like this:

```ts
const parent = event.composedPath().findLast(el => el instanceof PfeAccordion);
if (parent === this) {
  // collapse other panels
}
```

marionnegp commented:
Looks good to me now, @brianferry!

brianferry commented:
> @brianferry thinking it over i'm not certain we need context.
> 
> In the change handler, you could do something like this:
> 
> ```ts
> const parent = event.composedPath().findLast(el => el instanceof PfeAccordion);
> if (parent === this) {
>   // collapse other panels
> }
> ```

@bennypowers building off of this I was able to fix the issue now on https://deploy-preview-2114--patternfly-elements.netlify.app/components/accordion/demo/ 

The nested accordion is the first example on the demo currently and shows the nested accordion also with the single flag turned on.  Only real difference was using `find` instead of `findLast` and sending the parent into the function where we collapse the panels.

bennypowers commented:
will it work if you nest them 3 levels deep?

brianferry commented on .changeset/tough-days-unite.md:
```diff
@@ -0,0 +1,5 @@
+"@patternfly/pfe-accordion": minor
+---
+
+Moving this component to pfe-accordion 1:1
```
I have updated this to include more details / two examples (one reg & one nested accordion)

brianferry commented on elements/pfe-accordion/BaseAccordion.scss:
```diff
@@ -0,0 +1,24 @@
+
+:host([bordered]) ::slotted(pfe-accordion-header:first-child) {
+  display: block;
+  border-top: 1px solid #d2d2d2;
```
Updated & moved this into the pfe-accordion scss since that's where the attributes are also being passed in

brianferry commented on elements/pfe-accordion/BaseAccordionHeader.ts:
```diff
@@ -0,0 +1,168 @@
+
+  @query('.toggle') button?: HTMLButtonElement;
+
+  private _generatedHtag?: HTMLHeadingElement;
```
Yup, updated the ones that I could into the #private fields for these, some remain with the private tag like the initializer due to access restrictions though.

brianferry commented on elements/pfe-accordion/BaseAccordionHeader.ts:
```diff
@@ -0,0 +1,168 @@
+
+  @property({ attribute: 'aria-controls', reflect: true }) ariaControls?: string;
+
+  @property({ type: String, reflect: true }) bordered?: 'true'|'false';
```
Fair point, updated

brianferry commented:
> <pfe-band id="context-band" class="contextual" color-palette="lightest" full-width aside-height="full" aside-mobile="top">

It does, I've updated the demo so the last example shows all the nested accordions & three levels deep

bennypowers commented:
some notes from system fed discovery:
- we'd prefer to _remove_ all the url state code from this component, so that it becomes an app-level concern instead
- similarly, `expanded-index` attr should be removed in favour of `expanded` attr on individual panels. the js property can stay though, and so could any sort of `expandIndex(i: number): void` method

bennypowers commented on elements/pfe-accordion/BaseAccordionHeader.scss:
```diff
@@ -0,0 +1,113 @@
+:host {
+  --pf-c-accordion__toggle--Color: var(--pf-global--Color--100, #151515);
```
@markcaron do you know if these vars are meant to be user accessible? Is it ok if the user's preferences for these values from higher up the DOM tree are ignored?

bennypowers replied:
every instance of properties set on :host should be removed and set as fallbacks

bennypowers commented on elements/pfe-accordion/BaseAccordionHeader.scss:
```diff
@@ -0,0 +1,113 @@
+}
+
+.toggle[aria-expanded="true"] .icon {
+  transform: rotate(var(--pf-c-accordion__toggle--expanded-icon--Rotate, 90deg));
```
I think we should already prefer transform properties:

```suggestion
  rotate: var(--pf-c-accordion__toggle--expanded-icon--Rotate, 90deg);
```

marionnegp commented on elements/pfe-accordion/BaseAccordionHeader.scss:
```diff
@@ -0,0 +1,113 @@
+  --pf-c-accordion__toggle--PaddingBottom: var(--pf-global--spacer--sm, 0.5rem);
+  --pf-c-accordion__toggle--PaddingLeft: var(--pf-global--spacer--md, 1rem);
+  --pf-c-accordion__toggle--FontSize: var(--pf-global--FontSize--lg, 1rem);
+  --pf-c-accordion__toggle--FontFamily: var(--pf-global--FontFamily--redhat-updated--heading--sans-serif, "RedHatDisplayUpdated", helvetica, arial, sans-serif);
```
PF's using RedHatTextUpdated for their toggle headings. Sorry if I didn't flag this before!

brianferry replied:
No worries at all, I missed this as well.  Thank you!

marionnegp reviewed (changes requested):
Added one comment. The nested accordions look good!

bennypowers commented on elements/pfe-accordion/BaseAccordionHeader.scss:
```diff
@@ -0,0 +1,89 @@
+  --pfe-icon--size: var(--pf-c-accordion__toggle--IconSize, 10px);
+
+  color: var(--pf-c-accordion__toggle--Color, var(--pf-global--Color--100, #151515));
+  background-color: var(--pf-global--BackgroundColor--100);
```
```suggestion
  background-color: var(--pf-global--BackgroundColor--100, #ffffff);
```
